src/main/java/org/jenkins/plugins/lockableresources/LockStep.java

```diff
@@ -11,25 +11,47 @@
 import hudson.Extension;
 import hudson.model.AutoCompletionCandidates;
 
+import edu.umd.cs.findbugs.annotations.Nullable;
+import edu.umd.cs.findbugs.annotations.CheckForNull;
+
 public class LockStep extends AbstractStepImpl implements Serializable {
 
-	public final String resource;
+	@CheckForNull
+	public String resource = null;
+
+	@CheckForNull
+	public String label = null;
+
+	public int quantity = 0;
 
 	public boolean inversePrecedence = false;
 
 	@DataBoundConstructor
-	public LockStep(String resource) {
-		if (resource == null || resource.isEmpty()) {
-			throw new IllegalArgumentException("must specify resource");
-		}
-		this.resource = resource;
+	public LockStep() {
```

Comment:
You are right that this breaks backward compatibility.
But only because the first pipeline implementation done by cloudbees only focused on the usage of resource names (and skipped the usage of labels and quantities all together).

One could argue that either
- the first pipeline implementation by cloudbees shouldn't have been merged at all, because it focused only on 33% and did not provide an upgrade path (the problem you are looking at now)
- that increasing the functionality for the pipeline implementation to the full feature set of the classic non-pipeline implementation is allowed to introduce breaking-changes since pipelines are still under heavy development.

Hence, I don't agree with your assumption that due to a incomplete (wrong) implementation done previously, we are not able to correct this implementation and do it right.
Sometimes a breaking change must happen. The sooner the better.